#+TITLE: Repository for RNA-seq and microarray pipelines
#+AUTHOR: Sleiman Bassim, PhD
#+EMAIL: slei.bass@gmail.com

#+STARTUP: content
#+STARTUP: hidestars
#+OPTIONS: toc:5 H:5 num:3
#+LANGUAGE: english
#+LaTeX_HEADER: \usepackage[ttscale=.875]{libertine}
#+LATEX_HEADER: \usepackage[T1]{fontenc}
#+LaTeX_HEADER: \sectionfont{\normalfont\scshape}
#+LaTeX_HEADER: \subsectionfont{\normalfont\itshape}
#+LATEX_HEADER: \usepackage[innermargin=1.5cm,outermargin=1.25cm,vmargin=3cm]{geometry}
#+LATEX_HEADER: \linespread{1}
#+LATEX_HEADER: \setlength{\itemsep}{-30pt}
#+LATEX_HEADER: \setlength{\parskip}{0pt}
#+LATEX_HEADER: \setlength{\parsep}{-5pt}
#+LATEX_HEADER: \usepackage[hyperref]{xcolor}
#+LATEX_HEADER: \usepackage[colorlinks=true,urlcolor=SteelBlue4,linkcolor=Firebrick4]{hyperref}
#+EXPORT_SELECT_TAGS: export
#+EXPORT_EXCLUDE_TAGS: noexport

This git repository includes RNA-seq data analysis pipelines,
protocols, strategies, figures, and supplemental information
of peer-reviewed manuscripts. 

Licence: MIT Licence

* Language prerequisites
Shell/bash, Python, R, MySQL, D3.js

* Genetic analyses and published projects
1. [[https://github.com/neocruiser/pipelines][Repository of automated sequence analyses, statistics, and machine
   learning on high performance clusters]] (Bassim 2013-2018)
2. [[https://github.com/neocruiser/Rstats/tree/master/relapse][Supervised learning with gene expression data]] including multiclustering and genetic networks at the University Health Network (Bassim 2018) [[https://github.com/neocruiser/Rstats/blob/master/relapse/relapse.pdf][PDF]]
3. [[https://github.com/neocruiser/Rstats/tree/master/nodule][Networks of a cross-talk between host and pathogenic parasite with
   dual RNA-seq]] at the New York University (Bassim 2017) [[https://github.com/neocruiser/Rstats/blob/master/nodule/nodule.pdf][PDF]]
4. [[https://github.com/neocruiser/Rstats/tree/master/ganglia][Difference in networks between neural and respiratory cells with
   RNA-seq data]] at the New York University (Bassim 2016) [[https://github.com/neocruiser/Rstats/blob/master/ganglia/ganglia.pdf][PDF]]
5. [[https://github.com/neocruiser/Rstats/tree/master/rnaseQPX][Hotspots for variants in pathogenic parasite with RNA-seq data]] at
   the New York University (Bassim 2015) [[https://github.com/neocruiser/Rstats/blob/master/rnaseQPX/rnaseQPX.pdf][PDF]]
6. [[https://github.com/neocruiser/Rstats/tree/master/nodule#gene-gene-interaction][Functional domain curated databases for gene annotation]]
   (Bassim 2015) best used with [[https://github.com/neocruiser/pipelines/blob/master/mining/automated_analyses.sh][the automated aggregation of gene
   metadata script]] (Bassim 2017)
7. [[https://github.com/neocruiser/Rstats/blob/master/algae/algae.pdf][Descriptive analysis between algae species]] at the New York
   University (Bassim 2015)
8. [[https://github.com/neocruiser/Rstats/blob/master/brain.org#2-plink][Variant annotation and frequency in Parkinson's disease]] from
   clinical data at the University of Vermont (Bassim 2015) [[https://github.com/neocruiser/Rstats/blob/master/humanR/PD/parkinsons.pdf][PDF]]
9. [[https://github.com/neocruiser/Rstats/blob/master/brain.org][Brain disease repository including databases and workflows]] at the
   University of Vermont (Bassim 2015)
10. [[https://github.com/neocruiser/bibliographies][Aggregated bibliographies]] in machine learning, deep learning,
    brain disease methods, and non-model species bioinformatics
11. [[https://github.com/neocruiser/thesis2014/tree/master][Bassim PhD thesis]] involving [[https://github.com/neocruiser/thesis2014/blob/master/microarrays/preProcessing_detailed.R][microarray analyses]], [[https://github.com/neocruiser/thesis2014/blob/master/ebdbn/ebdbn.R][Bayesian
    networks]], [[https://github.com/neocruiser/thesis2014/blob/master/mfuzz/mfuzz.R][clustering]], and [[https://github.com/neocruiser/thesis2014/blob/master/Paper3/paper3.R][machine learning]] at the University of
    Quebec (Bassim 2014)

* Relevant material

| Subject              | Repo     | Code       | Author    | Year |
|----------------------+----------+------------+-----------+------|
| Statistical learning | [[https://github.com/neocruiser/Rstats/blob/master/linux.org#statistical-learning-summary][Summary]]  |            | S. Bassim | 2013 |
| Blast+               |          | [[https://github.com/neocruiser/Rstats/blob/master/linux.org#blast][shell]]      | S. Bassim | 2015 |
| Linux special addons | [[https://github.com/neocruiser/Rstats/blob/master/linux.org#additional-linux-apps][Table]]    |            | S. Bassim |      |
| Linux configurations | [[https://github.com/neocruiser/Rstats/blob/master/linux.org#collection-of-information][Software]] |            | S. Bassim | 2015 |
| D3.js summary        |          | [[https://github.com/neocruiser/Rstats/blob/master/linux.org#g3js-dataviz][javascript]] | S. Bassim |      |
| R summary functions  | [[https://github.com/neocruiser/Rstats/blob/master/linux.org#r-short-summary][Table]]    |            | S. Bassim |      |
| Emacs key-bindings   | [[https://github.com/neocruiser/Rstats/blob/master/linux.org#emacs-keybindings][Table]]    |            | S. Bassim |      |
| Archlinux configs    |          | [[https://github.com/neocruiser/Rstats/blob/master/linux.org#archlinux][shell]]      | S. Bassim |      |


* Author mentions
- Empirical Bayes Estimation of Dynamic Bayesian Networks (ebdbNet) on CRAN [[https://cran.r-project.org/web/packages/ebdbNet/NEWS][(updates]]).
- Google Scholar [[https://scholar.google.com/citations?user=mDJZY3oAAAAJ&hl=en][personal page]] for published papers.
- Linkedin [[https://www.linkedin.com/profile/view?id=AAMAAAD-CVIBL2ksOX7KT3k7ZUYPiE7A5XB7QFA&trk=hp-identity-name][profile]] for ongoing projects.
- Research gate [[https://www.researchgate.net/profile/Sleiman_Bassim][personal page]] for downloading published papers.
- Discovery of new PPAR target genes in the glioblastoma cell line T98G [[https://www.researchgate.net/profile/Stephane_Gibaud/publication/235963296_Development_of_microemulsion_of_mitotane_for_improvement_of_oral_bioavailability/links/0deec51fbd6341f6a2000000.pdf][Oncology]]
